### PR TITLE
Returning null values for user and campaigns properties not set

### LINKED ISSUE #155: Update user campaigns response object to return all properties
Off the top of my head, I think it's just...
```
{
  signup_id,
  signup_source,
  signup_group,
  reportback_id,
  campaign_id
}
```

And if there is no value set, it should return `null`.

Ref: https://github.com/DoSomething/northstar/issues/153

### PR DESCRIPTION
**Pull request by jonuy**
Sorta feels hacky? But other than modifying all documents already in the db to store null values for properties, there doesn't seem to be a better way to handle this other than with middleware.

#### What's this PR do?

Modifies the responses that include user and campaign objects to return all possible properties for them even if they're not set. Values default to null.

This is all the result of discussion that happened here: https://github.com/DoSomething/northstar/issues/153

Endpoints affected:

- `GET user/campaigns/{campaign_id}`
- `POST user/campaigns/{campaign_id}/signup`
- `POST user/campaigns/{campaign_id}/reportback`
- `PUT user/campaigns/{campaign_id}/reportback`
- `GET users/{term}/{id}`
- `POST users/{id}/avatar`
- `GET users/{term}/{id}/campaigns`

#### Where should the reviewer start?
The majority of the changes are in the new middleware files. Basically it's intercepting the response before it goes back to the user and fills in User and Campaign properties with null values for keys that aren't set.

#### Any background context?
https://github.com/DoSomething/northstar/issues/153

#### What are the relevant tickets?
Closes #155

### CODE REVIEW

**Comment by angaither** [comment]
:+1: looks ok, I would love if @DFurnes could give any additional middleware feelings, I haven't used it yet!

File: app/Http/Middleware/CampaignResponseMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+class CampaignResponseMiddleware {
+
+    public function handle($request, Closure $next)
+    {
+        $response = $next($request);
+        $response = $response->getData();
+
+        if (is_array($response->data)) {
+            foreach ($response->data as $campaign) {
+                self::fillCampaign($campaign);
```
Comment (DFurnes) [inline app/Http/Middleware/CampaignResponseMiddleware.php]: Should this be `$this->fillCampaign($campaign)` since the method isn't static?

**Comment by DFurnes** [comment]
Seems okay. I wonder if it might make more sense to add this as responsibility of the `respond()` base controller method? I feel like it seems kinda hidden away as a middleware.

File: app/Http/Middleware/CampaignResponseMiddleware.php (hunk 1 above)
Comment (jonuy) [inline app/Http/Middleware/CampaignResponseMiddleware.php]: oh yup, you right

**Comment by jonuy** [comment]
@DFurnes I thought about putting this in the base controller, but it felt meh because auth, signup-group, kudos, and api key controller code goes through that base controller too. With the middleware, I could at least pick and choose which routes to apply this to.

**Comment by DFurnes** [comment]
Fair, although I could see us wanting to do similar things for those items as well one day & having to make a middleware for each type of object could get tedious.

**Comment by jonuy** [comment]
True. Placing the logic in a single location in the base controller can make it less tedious.

We would however I think lose the decoupling of resources that are otherwise unrelated - the base controller being made aware of and working on the User, Campaign, ApiKey, and Token models.

Hm, even in the two middleware scenarios here, the code between the two, although pretty similar, still has slight variations in its logic. Something about keeping things decoupled still feels like the route to go here... but I can be swayed.

**Comment by DFurnes** [comment]
Hmm, that's fair. The nested resources do make it more complicated.

Seems fine to merge for now & just keep an eye on maybe? :+1: from me.

**Pull request merged**